File: lib/agents/positioning.ts
import { searchWeb, searchAdsTransparency } from '../tools/serpapi';
import { scrapePage } from '../tools/firecrawl';
import { searchReddit } from '../tools/reddit';
import { planQueries } from '../tools/query-planner';
import { filterRelevant, requirementsFromContext } from '../tools/relevance';
import { extractCandidates, verifyCandidates } from '../tools/candidate-discovery';
import { generateHuggingFaceJson } from './gemini';
import type {
  AgentConfig,
  AgentContext,
  AgentOutput,
  PositioningOutput,
  MessagingGap,
  AgentSource,
  ConfidenceLevel,
  EvidenceCandidate,
} from './types';
import { scoreToLevel } from './types';
import { computeSignalQualityPenalty, extractToolResults } from '../tools/fallback';
import {
  competitorSiteUrl,
  isPlaceholderCompetitor,
  isUsableScrapePage,
  productSiteUrl,
  skippedScrapePromise,
} from './entity-url';
import { runResearchLoop } from './research-loop';
import { applyInsufficientGate, evidencePromptRules } from './evidence-gate';
import { localeFromGeography } from './search-locale';
import {
  resolveEntityProbes,
  resolveGapQueries,
  resolveSearchQueries,
} from './plan-queries';
import type { SearchResult, ScrapedPage, ToolResult } from '../tools/types';
import type { discoverAndScrape } from '../tools/discover-and-scrape';

async function run(ctx: AgentContext): Promise<AgentOutput> {
  const { query, product, competitor, priorContext, geography, category, namedEntities, requiredTerms } = ctx;

  const competitorName =
    competitor && !isPlaceholderCompetitor(competitor) ? competitor : 'discovered competitors';
  const compUrl = competitorSiteUrl(ctx);
  const prodUrl = productSiteUrl(ctx);
  const locale = localeFromGeography(geography);
  const requirements = requirementsFromContext(ctx);

  const queryBundle = planQueries({
    product,
    competitor,
    domain: 'positioning',
    query,
    category,
    geography,
    namedEntities,
    requiredTerms,
  });

  const planned = resolveSearchQueries(ctx, [queryBundle.broad, queryBundle.targeted], 2);
  const primaryQuery = planned[0] ?? queryBundle.broad;
  const secondaryQuery = planned[1] ?? queryBundle.hypothesis;
  const entityProbes = resolveEntityProbes(ctx, queryBundle.entityProbes, 2);
  const gapQueries = resolveGapQueries(ctx, 3);

  const searchedFor = [
    primaryQuery,
    secondaryQuery,
    queryBundle.hypothesis,
    ...entityProbes,
    ...gapQueries,
  ];
  let candidates: EvidenceCandidate[] = [];

  const loop = await runResearchLoop({
    domain: 'positioning',
    requirements,
    budgetMs: 15_000,
    round1: () =>
      Promise.allSettled([
        compUrl ? scrapePage(compUrl) : skippedScrapePromise(),
        prodUrl ? scrapePage(prodUrl) : skippedScrapePromise(),
        isPlaceholderCompetitor(competitor)
          ? Promise.resolve({ data: [], timestamp: new Date().toISOString(), status: 'failed' as const, source: 'skip' })
          : searchAdsTransparency(competitorName),
        searchAdsTransparency(product),
        searchWeb(primaryQuery, locale),
        searchReddit(queryBundle.hypothesis),
        searchWeb(secondaryQuery, locale),
        ...entityProbes.slice(0, 1).map(q => searchWeb(q, locale)),
      ]),
    ingest: (settled, round) => {
      const sources: AgentSource[] = [];
      const rawContent: string[] = [];
      const allHits: SearchResult[] = [];

      const takeSearch = (
        result: PromiseSettledResult<unknown> | undefined,
        label: string,
        tool: AgentSource['tool'],
        limit: number,
      ) => {
        if (!result || result.status !== 'fulfilled') return;
        const value = result.value as ToolResult<SearchResult[]>;
        if (!Array.isArray(value?.data)) return;
        const asSearch = value.data.map(r => ({
          title: r.title,
          url: r.url,
          snippet: 'snippet' in r ? String((r as SearchResult).snippet ?? '') : '',
        }));
        const { kept } = filterRelevant(asSearch, requirements, { limit, minScore: geography ? 0.25 : 0.15 });
        kept.forEach(r => {
          allHits.push(r);
          sources.push({ url: r.url, title: r.title, timestamp: value.timestamp, tool });
          rawContent.push(`[${label}] ${r.title}: ${r.snippet}`);
        });
      };

      if (round === 1) {
        if (settled[0] && isUsableScrapePage(settled[0] as PromiseSettledResult<ToolResult<ScrapedPage>>)) {
          const page = (settled[0] as PromiseFulfilledResult<ToolResult<ScrapedPage>>).value.data;
          sources.push({ url: page.url, title: competitor ? `${competitor} — homepage` : 'Competitor homepage', timestamp: (settled[0] as PromiseFulfilledResult<ToolResult<ScrapedPage>>).value.timestamp, tool: 'firecrawl' });
          rawContent.push(`[COMPETITOR HOMEPAGE] ${page.excerpt}`);
        }
        if (settled[1] && isUsableScrapePage(settled[1] as PromiseSettledResult<ToolResult<ScrapedPage>>)) {
          const page = (settled[1] as PromiseFulfilledResult<ToolResult<ScrapedPage>>).value.data;
          sources.push({ url: page.url, title: product.length < 50 ? `${product} — homepage` : 'Product homepage', timestamp: (settled[1] as PromiseFulfilledResult<ToolResult<ScrapedPage>>).value.timestamp, tool: 'firecrawl' });
          rawContent.push(`[OUR HOMEPAGE] ${page.excerpt}`);
        }
        takeSearch(settled[2], 'COMPETITOR AD', 'serpapi', 3);
        takeSearch(settled[3], 'OUR AD', 'serpapi', 3);
        takeSearch(settled[4], 'MESSAGING SEARCH', 'serpapi', 4);
        takeSearch(settled[5], 'REDDIT PERCEPTION', 'reddit', 3);
        takeSearch(settled[6], 'SOCIAL VOICE', 'serpapi', 3);
        for (let i = 7; i < settled.length; i++) takeSearch(settled[i], 'ENTITY POSITIONING', 'serpapi', 3);
        candidates = extractCandidates(allHits, { geographyName: geography?.name, exclude: [product, competitorName], limit: 5 });
      } else {
        for (const s of settled) {
          if (s.status !== 'fulfilled') continue;
          const val = s.value as Awaited<ReturnType<typeof discoverAndScrape>> | ToolResult<SearchResult[]>;
          if (val && 'pages' in val && Array.isArray(val.pages)) {
            val.pages.forEach(pageResult => {
              if (pageResult.status === 'failed' || !pageResult.data.markdown?.trim()) return;
              sources.push({ url: pageResult.data.url, title: pageResult.data.title || 'Candidate', timestamp: pageResult.timestamp, tool: 'firecrawl' });
              rawContent.push(`[CANDIDATE SITE] ${pageResult.data.title}: ${pageResult.data.excerpt}`);
            });
          } else if (val && 'data' in val && Array.isArray(val.data)) {
            takeSearch(
              { status: 'fulfilled', value: val } as PromiseFulfilledResult<ToolResult<SearchResult[]>>,
              'GAP QUERY',
              'serpapi',
              4,
            );
          }
        }
      }

      return {
        sources,
        rawContent,
        toolResults: extractToolResults(settled as PromiseSettledResult<ToolResult<unknown>>[]),
        searchedFor,
        relevantSourceCount: sources.length,
        relevantHits: allHits,
        scrapedPageCount: sources.filter(s => s.tool === 'firecrawl').length,
      };
    },
    gapRound: async () => {
      if (gapQueries.length > 0) {
        const settled = await Promise.allSettled(
          gapQueries.slice(0, 2).map(q => searchWeb(q, locale)),
        );
        searchedFor.push(...gapQueries.slice(0, 2));
        return settled;
      }

      const { settled, queries } = await verifyCandidates(candidates, {
        product,
        geographyName: geography?.name,
        category: category ?? 'positioning',
        maxCandidates: 2,
        topN: 1,
      });
      searchedFor.push(...queries);
      return settled;
    },
  });

  if (candidates.length) loop.evidence = { ...loop.evidence, candidates };

  const systemPrompt = `You are a brand positioning strategist. Positioning is how you talk about what exists.
${priorContext ? `\nPrior conversation context:\n${priorContext}` : ''}
${evidencePromptRules(geography, category)}`;

  const userPrompt = `Query: "${query}"
Our product: ${product}
Competitor: ${competitorName}
${geography ? `Geography: ${geography.name}` : ''}
Evidence status: ${loop.evidence.status}

Raw signals:
${loop.rawContent.join('\n') || '(no relevant positioning signals)'}

Produce JSON:
{
  "insufficientEvidence": boolean,
  "facts": string[],
  "interpretation": string[],
  "yourPositioning": string,
  "competitorPositioning": string,
  "gaps": [{ "dimension": string, "yourMessage": string, "competitorMessage": string, "gap": string, "opportunity": string }],
  "adThemes": string[],
  "synthesizedAnswer": string,
  "confidenceScore": number
}`;

  let parsed: Record<string, unknown> = {};
  try {
    parsed = await generateHuggingFaceJson(systemPrompt, userPrompt, { maxNewTokens: 1400, temperature: 0.2 });
  } catch {
    parsed = { insufficientEvidence: true, facts: [], interpretation: loop.evidence.gaps, gaps: [], adThemes: [], confidenceScore: 0.35 };
  }

  const gate = applyInsufficientGate(parsed, loop.evidence);
  if (gate.insufficient) loop.evidence = { ...loop.evidence, status: 'insufficient' };

  const confScore = Number.parseFloat(
    (gate.confidenceScore * computeSignalQualityPenalty(loop.toolResults, 7)).toFixed(2),
  );

  const output: PositioningOutput = {
    agentId: 'positioning',
    domain: 'positioning',
    artifactType: 'positioning-gap',
    confidence: scoreToLevel(confScore),
    confidenceScore: confScore,
    facts: gate.insufficient ? [] : (parsed.facts as string[] ?? []),
    interpretation: gate.insufficient ? loop.evidence.gaps : (parsed.interpretation as string[] ?? []),
    sources: loop.sources,
    generatedAt: new Date().toISOString(),
    competitor: competitorName,
    yourPositioning: gate.insufficient ? '' : (parsed.yourPositioning as string) ?? '',
    competitorPositioning: gate.insufficient ? '' : (parsed.competitorPositioning as string) ?? '',
    gaps: gate.insufficient ? [] : ((parsed.gaps ?? []) as MessagingGap[]),
    adThemes: gate.insufficient ? [] : (parsed.adThemes as string[] ?? []),
    evidence: loop.evidence,
    toolCallCount: loop.toolCallCount,
    searchCallCount: loop.searchCallCount,
    scrapeCallCount: loop.scrapeCallCount,
    droppedIrrelevantCount: loop.droppedIrrelevantCount,
  };

  return output;
}

export const positioningAgent: AgentConfig = {
  id: 'positioning',
  name: 'Positioning Agent',
  description: 'Analyses homepages, ads, and marketing copy to surface messaging gaps and positioning opportunities.',
  run,
};

File: backend/app/agents/positioning.py
"""Positioning & Messaging Gaps agent (parity with lib/agents/positioning.ts)."""

from __future__ import annotations

import asyncio
import re
from datetime import datetime, timezone
from typing import Any

from app.agents.entity_url import (
    competitor_site_url,
    is_usable_scrape_page,
    product_site_url,
    skipped_scrape_promise,
)
from app.gemini import generate_json
from app.models import score_to_level
from app.tools.fallback import compute_signal_quality_penalty, extract_tool_results
from app.tools.firecrawl import scrape_page
from app.tools.reddit import search_reddit
from app.tools.serpapi import search_ads_transparency, search_web

AGENT_ID = "positioning"
AGENT_NAME = "Positioning Agent"
AGENT_DESCRIPTION = (
    "Analyses homepages, ads, and marketing copy to surface messaging gaps and positioning opportunities."
)


def _fulfilled(result: Any) -> bool:
    return not isinstance(result, BaseException)


async def run(ctx: dict[str, Any]) -> dict[str, Any]:
    query = ctx["query"]
    product = ctx["product"]
    competitor = ctx.get("competitor")
    prior_context = ctx.get("prior_context")

    competitor_name = competitor or "relevant competitors"
    comp_url = competitor_site_url(ctx)
    prod_url = product_site_url(ctx)

    (
        comp_home_result,
        prod_home_result,
        comp_ads_result,
        prod_ads_result,
        messaging_search_result,
        reddit_perception_result,
        social_voice_result,
    ) = await asyncio.gather(
        scrape_page(comp_url) if comp_url else skipped_scrape_promise(),
        scrape_page(prod_url) if prod_url else skipped_scrape_promise(),
        search_ads_transparency(competitor_name),
        search_ads_transparency(product),
        search_web(f"{competitor_name} vs {product} messaging positioning marketing"),
        search_reddit(f"how does {competitor_name} market itself brand positioning"),
        search_web(
            f"{competitor_name} OR {product} "
            "site:x.com OR site:twitter.com OR site:instagram.com OR site:linkedin.com positioning messaging"
        ),
        return_exceptions=True,
    )

    comp_about_url = f"{comp_url.rstrip('/')}/about" if comp_url else ""
    prod_about_url = f"{prod_url.rstrip('/')}/about" if prod_url else ""

    comp_about_result, prod_about_result = await asyncio.gather(
        scrape_page(comp_about_url) if comp_about_url else skipped_scrape_promise(),
        scrape_page(prod_about_url) if prod_about_url else skipped_scrape_promise(),
        return_exceptions=True,
    )

    sources: list[dict[str, Any]] = []
    raw_content: list[str] = []

    if is_usable_scrape_page(comp_home_result):
        page = comp_home_result["data"]
        title = f"{competitor} — homepage" if competitor else "Competitor homepage"
        sources.append(
            {
                "url": page["url"],
                "title": title,
                "timestamp": comp_home_result["timestamp"],
                "tool": "firecrawl",
            }
        )
        raw_content.append(f"[COMPETITOR HOMEPAGE] {page['excerpt']}")

    if is_usable_scrape_page(prod_home_result):
        page = prod_home_result["data"]
        title = f"{product} — homepage" if len(product) < 50 else "Product homepage"
        sources.append(
            {
                "url": page["url"],
                "title": title,
                "timestamp": prod_home_result["timestamp"],
                "tool": "firecrawl",
            }
        )
        raw_content.append(f"[OUR HOMEPAGE] {page['excerpt']}")

    if is_usable_scrape_page(comp_about_result):
        raw_content.append(f"[COMPETITOR ABOUT] {comp_about_result['data']['excerpt']}")

    if is_usable_scrape_page(prod_about_result):
        raw_content.append(f"[OUR ABOUT] {prod_about_result['data']['excerpt']}")

    if _fulfilled(comp_ads_result):
        for r in (comp_ads_result.get("data") or [])[:3]:
            sources.append(
                {
                    "url": r["url"],
                    "title": r["title"],
                    "timestamp": comp_ads_result["timestamp"],
                    "tool": "serpapi",
                }
            )
            raw_content.append(f"[COMPETITOR AD] {r['title']}: {r['snippet']}")

    if _fulfilled(prod_ads_result):
        for r in (prod_ads_result.get("data") or [])[:3]:
            sources.append(
                {
                    "url": r["url"],
                    "title": r["title"],
                    "timestamp": prod_ads_result["timestamp"],
                    "tool": "serpapi",
                }
            )
            raw_content.append(f"[OUR AD] {r['title']}: {r['snippet']}")

    if _fulfilled(messaging_search_result):
        for r in (messaging_search_result.get("data") or [])[:4]:
            sources.append(
                {
                    "url": r["url"],
                    "title": r["title"],
                    "timestamp": messaging_search_result["timestamp"],
                    "tool": "serpapi",
                }
            )
            raw_content.append(f"[MESSAGING SEARCH] {r['title']}: {r['snippet']}")

    if _fulfilled(reddit_perception_result):
        for p in (reddit_perception_result.get("data") or [])[:3]:
            sources.append(
                {
                    "url": p["url"],
                    "title": p["title"],
                    "timestamp": p["created"],
                    "tool": "reddit",
                }
            )
            raw_content.append(f"[REDDIT PERCEPTION] {p['title']}: {p['snippet']}")

    if _fulfilled(social_voice_result):
        for r in (social_voice_result.get("data") or [])[:3]:
            sources.append(
                {
                    "url": r["url"],
                    "title": r["title"],
                    "timestamp": social_voice_result["timestamp"],
                    "tool": "serpapi",
                }
            )
            raw_content.append(f"[SOCIAL VOICE] {r['title']}: {r['snippet']}")

    prior_block = f"\nPrior conversation context:\n{prior_context}" if prior_context else ""
    system_prompt = f"""You are a brand positioning strategist. You analyse how companies talk about themselves — their hero message, value frame, audience language — and find gaps and opportunities.

Key insight: Positioning is not what you build, it's how you talk about what already exists. A company can have the same product but win or lose based on messaging.

Look for:
- Value framing differences (technology-first vs outcome-first)
- Audience language differences
- Emotional vs functional emphasis
- Category claim differences (e.g. "AI SDR" vs "Revenue automation"){prior_block}"""

    user_prompt = f"""Query: "{query}"
Our product: {product}
Competitor: {competitor_name}

Raw signals:
{chr(10).join(raw_content)}

Produce JSON:
{{
  "facts": string[],
  "interpretation": string[],
  "yourPositioning": string,
  "competitorPositioning": string,
  "gaps": [
    {{
      "dimension": string,
      "yourMessage": string,
      "competitorMessage": string,
      "gap": string,
      "opportunity": string
    }}
  ],
  "adThemes": string[],
  "synthesizedAnswer": string,
  "confidenceScore": number
}}

Dimensions to analyse: Value Framing, Audience Language, Category Claim, Emotional Appeal, Social Proof Style, Feature Focus vs Outcome Focus, Brand Personality."""

    try:
        parsed = await generate_json(system_prompt, user_prompt, max_new_tokens=1400, temperature=0.2)
    except Exception:
        parsed = {
            "facts": [
                re.sub(r"^\[[^\]]+\]\s*", "", s)
                for s in raw_content[:3]
                if len(re.sub(r"^\[[^\]]+\]\s*", "", s)) > 15
            ],
            "interpretation": [
                "Analysis synthesis is temporarily unavailable. Raw data signals are shown below."
            ],
            "yourPositioning": "",
            "competitorPositioning": "",
            "gaps": [],
            "adThemes": [],
            "synthesizedAnswer": "Positioning data collected but synthesis failed.",
            "confidenceScore": 0.4,
        }

    raw_score = parsed["confidenceScore"] if isinstance(parsed.get("confidenceScore"), (int, float)) else 0.6
    tool_results = extract_tool_results(
        [
            comp_home_result,
            prod_home_result,
            comp_ads_result,
            prod_ads_result,
            messaging_search_result,
            reddit_perception_result,
            social_voice_result,
            comp_about_result,
            prod_about_result,
        ]
    )
    conf_score = round(float(raw_score) * compute_signal_quality_penalty(tool_results, 9), 2)
    confidence = score_to_level(conf_score)

    return {
        "agentId": AGENT_ID,
        "domain": "positioning",
        "artifactType": "positioning-gap",
        "confidence": confidence,
        "confidenceScore": conf_score,
        "facts": parsed.get("facts") or [],
        "interpretation": parsed.get("interpretation") or [],
        "sources": sources,
        "generatedAt": datetime.now(timezone.utc).isoformat(),
        "competitor": competitor_name,
        "yourPositioning": parsed.get("yourPositioning") or "",
        "competitorPositioning": parsed.get("competitorPositioning") or "",
        "gaps": parsed.get("gaps") or [],
        "adThemes": parsed.get("adThemes") or [],
    }
